pallets/did/src/lib.rs

```diff
@@ -28,91 +30,403 @@ mod tests;
 pub mod benchmarking;
 
 pub mod default_weights;
-pub mod migration;
 pub use default_weights::WeightInfo;
 
 use codec::{Decode, Encode};
-use frame_support::{decl_event, decl_module, decl_storage, dispatch::DispatchResult, Parameter, StorageMap};
+
+use codec::EncodeLike;
+use frame_support::{ensure, storage::types::StorageMap};
 use frame_system::{self, ensure_signed};
-use sp_runtime::{codec::Codec, traits::Member};
-use sp_std::prelude::*;
+use sp_core::{ed25519, sr25519};
+use sp_runtime::traits::Verify;
+use sp_std::{collections::btree_set::BTreeSet, convert::TryFrom, fmt::Debug, prelude::Clone, vec::Vec};
+
+pub use pallet::*;
+
+/// Reference to a payload of data of variable size.
+pub type PayloadReference<'a> = &'a [u8];
+
+/// Type of a signature (variable size as different signature schemes are
+/// supported).
+pub type SignatureEncoding = Vec<u8>;
 
-/// The DID trait
-pub trait Config: frame_system::Config {
-	/// DID specific event type
-	type Event: From<Event<Self>> + Into<<Self as frame_system::Config>::Event>;
+/// Reference to a signature of variable size.
+pub type SignatureReference<'a> = &'a [u8];
 
-	/// Weight information for extrinsics in this pallet.
-	type WeightInfo: WeightInfo;
+/// Type for an encoded URL.
+pub type UrlEncoding = Vec<u8>;
 
-	/// Public signing key type for DIDs
-	type PublicSigningKey: Parameter + Member + Codec + Default;
+/// Trait representing a public key under the control of a DID subject.
+pub trait DIDPublicKey {
+	/// Returns the key method description as in the [DID specification](https://w3c.github.io/did-spec-registries/#verification-method-types).
+	fn get_did_key_description(&self) -> &'static str;
+}
+
+/// An enum describing the different verification methods a verification key can fulfil, according to the [DID specification](https://w3c.github.io/did-spec-registries/#verification-relationships).
+#[derive(Clone, Debug, Decode, Encode, PartialEq, Eq)]
+pub enum DIDVerificationKeyType {
+	Authentication,
+	CapabilityDelegation,
+	CapabilityInvocation, // Not used for now, but added for potential future use
+	AssertionMethod,
+}
+
+/// An enum describing the different verification methods an encryption key can fulfil, according to the [DID specification](https://w3c.github.io/did-spec-registries/#verification-relationships).
+#[derive(Clone, Debug, Decode, Encode, PartialEq)]
+pub enum DIDEncryptionKeyType {
+	KeyAgreement,
+}
 
-	/// Public boxing key type for DIDs
-	type PublicBoxKey: Parameter + Member + Codec + Default;
+/// Enum representing the types of verification keys a DID can control.
+#[derive(Clone, Decode, Debug, Encode, Eq, Ord, PartialEq, PartialOrd)]
+pub enum PublicVerificationKey {
+	/// An Ed25519 public key.
+	Ed25519([u8; 32]),
+	/// A Sr25519 public key.
+	Sr25519([u8; 32]),
```

Comment:
```suggestion
	Ed25519(ed25519::Public),
	/// A Sr25519 public key.
	Sr25519(sr25519::Public),
```

Reply:
You can also consider to use [MultiSignature](https://github.com/paritytech/substrate/blob/f16abdac4d39542f5eacc8041a7d1ce573278034/primitives/runtime/src/lib.rs#L237) and [MultiSigner](https://github.com/paritytech/substrate/blob/f16abdac4d39542f5eacc8041a7d1ce573278034/primitives/runtime/src/lib.rs#L294). But I would understand if you rather not use them because of ecdsa.